File: cli/src/main/java/io/bitken/ss/cli/plan/Tag.java
package io.bitken.ss.cli.plan;

import io.bitken.ss.cli.HasSpec;
import io.bitken.ss.gw.GitTags;
import picocli.CommandLine.Model.CommandSpec;
import picocli.CommandLine.Model.OptionSpec;

import java.util.Set;
import java.util.concurrent.Callable;

/**
 * {@code plan tag --plan N --kind version|complete|abandoned}
 *
 * <p>Creates the appropriate local git tag and prints the push line.
 * For {@code --kind version}: computes the next vK, refuses if it already
 * exists, creates it, and prints {@code git push origin plan-N-vK}.
 * For {@code complete} / {@code abandoned}: creates the fixed tag and prints
 * the push line. Does not call {@code git push}.
 */
public class Tag implements Callable<Integer>, HasSpec {

    private static final Set<String> FIXED_KINDS = Set.of("complete", "abandoned");

    private final CommandSpec spec;
    private final GitTags gitTags;

    public Tag(GitTags gitTags) {
        this.gitTags = gitTags;
        spec = CommandSpec.wrapWithoutInspection(this);
        spec.name("tag");
        spec.usageMessage().description("Create a plan version/complete/abandoned tag.");
        spec.addOption(OptionSpec.builder("--plan").required(true).type(int.class).build());
        spec.addOption(OptionSpec.builder("--kind").required(true).type(String.class).build());
    }

    @Override
    public CommandSpec getSpec() { return spec; }

    @Override
    public Integer call() {
        var pr = spec.commandLine().getParseResult();
        int plan = pr.matchedOption("plan").getValue();
        String kind = pr.matchedOption("kind").getValue();

        if ("version".equals(kind)) return createVersionTag(plan);
        if (FIXED_KINDS.contains(kind)) return createFixedTag(plan, kind);

        System.out.println("ERROR: --kind must be one of: version, complete, abandoned");
        return 1;
    }

    private int createVersionTag(int plan) {
        String tag = gitTags.nextPlanVersion(plan);
        if (gitTags.tagExists(tag)) {
            System.out.println("ERROR: tag " + tag + " already exists — commit more changes before re-tagging");
            return 1;
        }
        return createAndPrint(tag);
    }

    private int createFixedTag(int plan, String kind) {
        return createAndPrint("plan-" + plan + "-" + kind);
    }

    private int createAndPrint(String tag) {
        if (!gitTags.createTag(tag)) {
            System.out.println("ERROR: failed to create tag " + tag);
            return 1;
        }
        System.out.println("Created tag: " + tag);
        System.out.println("Run: git push origin " + tag);
        return 0;
    }
}

File: exp/rust/crates/ss-cli/src/plan/tag.rs
//! `plan tag --plan N --kind version|complete|abandoned`
//!
//! Port of the Java `Tag`. Creates the local git tag and prints the push
//! line; it never pushes. For `version` the next vK is derived from the git
//! tags themselves (`GitTags::next_plan_version`) and refused if it already
//! exists.
//!
//! Note: 02-cli.md recorded a defect here — "derives the version from the XML
//! field, not git tags". That is stale: the Java `Tag` never reads the XML
//! (verified against the source). There is no defect to preserve.

use ss_core::gw::GitTags;

const FIXED_KINDS: [&str; 2] = ["complete", "abandoned"];

pub fn run(git_tags: &GitTags, plan: u32, kind: &str) -> i32 {
    match kind {
        "version" => create_version_tag(git_tags, plan),
        k if FIXED_KINDS.contains(&k) => create_and_print(git_tags, &format!("plan-{plan}-{k}")),
        _ => {
            println!("ERROR: --kind must be one of: version, complete, abandoned");
            1
        }
    }
}

fn create_version_tag(git_tags: &GitTags, plan: u32) -> i32 {
    let tag = git_tags.next_plan_version(plan);
    // Defensive, and unreachable in practice: next_plan_version returns
    // highest + 1, so the tag it names cannot already exist. Java carries the
    // same guard, so it is ported rather than dropped — but nothing can test
    // it without reaching past the gateway.
    if git_tags.tag_exists(&tag) {
        println!("ERROR: tag {tag} already exists — commit more changes before re-tagging");
        return 1;
    }
    create_and_print(git_tags, &tag)
}

fn create_and_print(git_tags: &GitTags, tag: &str) -> i32 {
    if !git_tags.create_tag(tag) {
        println!("ERROR: failed to create tag {tag}");
        return 1;
    }
    println!("Created tag: {tag}");
    println!("Run: git push origin {tag}");
    0
}
